File: part2/listing_0065_haversine_formula.py
from cmath import cos, sin, sqrt, asin
from math import atan2


def degree_to_radian(degree: float) -> float:
    return 0.01745329251994329577 * degree


def reference_haversine(lon1: float, lat1: float, lon2: float, lat2: float, earth_radius: float = 6372.8) -> complex:
    dlat = degree_to_radian(lat2 - lat1)
    dlon = degree_to_radian(lon2 - lon1)
    lat1 = degree_to_radian(lat1)
    lat2 = degree_to_radian(lat2)

    a = (sin(dlat / 2.0)) ** 2 + cos(lat1) * cos(lat2) * (sin(dlon / 2.0)) ** 2
    c = 2.0 * asin(sqrt(a))

    return earth_radius * c

File: part2/rust-parser/src/listing_0065_haversine_formula.rs
const EARTH_RADIUS: f64 = 6372.8;

fn degree_to_radian(degree: f64) -> f64 {
    degree * 0.01745329251994329577
}

pub fn reference_haversine(lon1: f64, lat1: f64, lon2: f64, lat2: f64) -> f64 {
    let dlat = degree_to_radian(lat2 - lat1);
    let dlon = degree_to_radian(lon2 - lon1);
    let lat1 = degree_to_radian(lat1);
    let lat2 = degree_to_radian(lat2);

    let a = f64::powf((dlat / 2.0).sin(), 2.0) + lat1.cos() * lat2.cos() * f64::powf((dlon / 2.0).sin(), 2.0);
    let c = 2.0 * a.sqrt().asin();
    EARTH_RADIUS * c
}
